File: C++/2-Medium/00311._稀疏矩阵的乘法/00311._稀疏矩阵的乘法.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> multiply(vector<vector<int>>& mat1, vector<vector<int>>& mat2) {
    int rows1 = mat1.size(), cols1 = mat1[0].size();
    int rows2 = mat2.size(), cols2 = mat2[0].size();
    vector<vector<int>> result(rows1, vector<int>(cols2, 0));
    
    for (int i = 0; i < rows1; ++i) {
        for (int k = 0; k < cols1; ++k) {
            if (mat1[i][k] != 0) { // 只处理非零元素
                for (int j = 0; j < cols2; ++j) {
                    if (mat2[k][j] != 0) { // 寻找对应的非零元素
                        result[i][j] += mat1[i][k] * mat2[k][j]; // 计算乘积并累加
                    }
                }
            }
        }
    }
    return result;
}

int main() {
    vector<vector<int>> mat1 = {{1,0,0},{-1,0,3}};
    vector<vector<int>> mat2 = {{7,0,0},{0,0,0},{0,0,1}};
    vector<vector<int>> result = multiply(mat1, mat2);

    for (const auto &row : result) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << "\\n";
    }

    return 0;
}

File: Python/2-Medium/00311._稀疏矩阵的乘法/00311._稀疏矩阵的乘法.py
from typing import List

def sparse_matrix_multiplication(mat1: List[List[int]], mat2: List[List[int]]) -> List[List[int]]:
    """
    计算两个稀疏矩阵的乘积。

    参数:
    mat1 (List[List[int]]): 第一个稀疏矩阵。
    mat2 (List[List[int]]): 第二个稀疏矩阵。

    返回:
    List[List[int]]: 两个稀疏矩阵的乘积。

    示例:
    >>> sparse_matrix_multiplication([[1,0,0],[-1,0,3]], [[7,0,0],[0,0,0],[0,0,1]])
    [[7, 0, 0], [-7, 0, 3]]
    """

    # 获取矩阵的维度
    mat1_rows, mat1_cols = len(mat1), len(mat1[0])
    mat2_cols = len(mat2[0])

    # 初始化结果矩阵
    result = [[0] * mat2_cols for _ in range(mat1_rows)]

    # 遍历mat1的每一行
    for i in range(mat1_rows):
        # 对于mat1的每一列，如果该元素不为0，则进行计算
        for j in range(mat1_cols):
            if mat1[i][j] != 0:
                # 对mat2的每一列进行遍历，计算乘积并累加到结果矩阵对应位置
                for k in range(mat2_cols):
                    result[i][k] += mat1[i][j] * mat2[j][k]

    return result

# 测试代码
mat1 = [[1,0,0],[-1,0,3]]
mat2 = [[7,0,0],[0,0,0],[0,0,1]]
result = sparse_matrix_multiplication(mat1, mat2)
print(result)  # [[7, 0, 0], [-7, 0, 3]]